src/main/java/com/abrus/user/service/UserServiceImpl.java

```diff
@@ -0,0 +1,64 @@
+package com.abrus.user.service;
+
+import com.abrus.user.exception.ValidationException;
+import com.abrus.user.model.User;
+import com.abrus.user.model.UserRegistrationDto;
+import com.abrus.user.model.UserResponseDto;
+import com.abrus.user.repository.UserRepository;
+import com.abrus.user.validator.ValidationResult;
+import com.abrus.user.validator.UserRegistrationValidator;
+import lombok.NonNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class UserServiceImpl implements UserService {
+    @NonNull
+    private final UserRepository userRepository;
+    @NonNull
+    private final UserRegistrationValidator userRegistrationValidator;
+
+    public UserServiceImpl(UserRepository userRepository, UserRegistrationValidator userRegistrationValidator) {
+        this.userRepository = userRepository;
+        this.userRegistrationValidator = userRegistrationValidator;
+    }
+
+    @Override
+    public UserResponseDto registerUser(UserRegistrationDto userRegistrationDto) {
+        ValidationResult validationResult = userRegistrationValidator.validate(userRegistrationDto);
+        if (!validationResult.getIsValid()) {
+            throw new ValidationException(validationResult.getErrorMsg());
+        }
+        User user = new User();
+        user.setEmail(userRegistrationDto.email());
+        user.setPassword(userRegistrationDto.password());
+        user.setPhoneNumber(userRegistrationDto.phoneNumber());
+
+        User newUser = userRepository.insert(user);
+        Optional<User> storedUser = userRepository.getById(newUser.getId());
+        if (storedUser.isPresent()) {
+            return new UserResponseDto(storedUser.get().getEmail(), storedUser.get().getPhoneNumber());
+        }
+        throw new RuntimeException("New user is not found in db: " + newUser.toString());
+    }
+
+    @Override
+    public Optional<UserResponseDto> getUserById(Long userId) {
+        Optional<User> storedUser = userRepository.getById(userId);
+        if (storedUser.isPresent()) {
+            User user = storedUser.get();
+            return Optional.of(new UserResponseDto(user.getEmail(), user.getPhoneNumber()));
+        }
+        return Optional.empty();
+    }
```

Comment:
Use orElseThrow to throw a DataProcessingException if the user is not present in the repository. This is the most common way to implement such methods. 
```suggestion
    @Override
    public UserResponseDto getUserById(Long userId) {
        User user = userRepository.getById(userId)
            .orElseThrow(() -> 
                new DataProcessingException("User not found with ID: " + userId)
            );

        return new UserResponseDto(user.getEmail(), user.getPhoneNumber());
    }
```

Reply:
Of course, you still may return Optional. In this case, you can optimize your solution:
```suggestion
    @Override
public Optional<UserResponseDto> getUserById(Long userId) {
    return userRepository.getById(userId)
        .map(user -> new UserResponseDto(user.getEmail(), user.getPhoneNumber()));
}
```